A new PR to resolve the conflicts in #78.

tel3path opened the pull request:
#78 could not be reopened, but has conflicts that need to be resolved. This new PR does not introduce any changes beyond what should have been merged from #78 and #79.

tel3path commented:
bors try

ltratt commented:
This has got 18 commits, so something has gone wrong.

tel3path commented:
bors try

0152la commented:
Have there been any changes between this try and the previous one?

tel3path commented:
> If not, I personally don't see a reason for the try; if you have one though, please tell me and let it continue. (also cancelling won't do anything on the buildbot side for now, unfortunately, so it'll continue either way)

The only reason is I thought my connection might have been dropped when I did the force push.

0152la commented:
I don't see a force push from you in this PR. Which one are you referring to?

tel3path commented:
> 

That's exactly why I thought my connection was dropped. Because I don't see the force push either.

tel3path commented:
> 

None of my attempts at force push are showing here. 
I did:
`git rebase -i upstream/master`
and got
`Successfully rebased and updated refs/heads/compare_platforms.`
then:
`git push origin +compare_platforms:master`
and got
`Everything up-to-date`

And since that didn't show I tried 
`git push origin compare_platforms -f`
and got
`Everything up-to-date`

tel3path commented:
bors try

tel3path commented:
bors try

ltratt commented:
This is now starting to look about right. But before we actually merge this, I just realised that the C file is using tabs. Please convert all tabs to 4 spaces, please (as a new commit, not a force push).

0152la commented:
I thought indenting should be covered by `clang-format`. We could amend the options we use to check for formatting.

ltratt commented:
It may or may not be fussy about tabs. Dunno. Bitter experience has taught me to be fussy about tabs, however!

tel3path commented:
> 

OK. I promise you my editor is using spaces and not tabs. Will convert now.

0152la commented:
Sure, what I'm saying is `clang-format` might be forcing tabs. And I think it is, perusing the repo. If we want spaces, we'd need to amend the `clang-format` check appropriately in the CI script.

ltratt commented:
Erp, OK, I stand correcte!. We should fix the clang-format settings, but not for this PR. My apologies -- please ignore what I said!

tel3path commented:
> Erp, OK, I stand correcte!. We should fix the clang-format settings, but not for this PR. My apologies -- please ignore what I said!

OK

tel3path commented:
> 

This, below:

> buildbot/ci-builder Expected — Waiting for status to be reported

I haven't seen this before. 

It also says

> Required statuses must pass before merging
All required [statuses](https://docs.github.com/rest/commits/statuses) and check runs on this pull request must run successfully to enable automatic merging.

Is this a step that bors will complete, or is it something we need to fix?

ltratt commented:
@0152la I think this now looks in good order so will leave it to you to shepherd through.

tel3path commented:
> This is now starting to look about right. But before we actually merge this, I just realised that the C file is using tabs. Please convert all tabs to 4 spaces, please (as a new commit, not a force push).

> 

This, below:

> buildbot/ci-builder Expected — Waiting for status to be reported

I haven't seen this before. 

It also says

> Required statuses must pass before merging
All required [statuses](https://docs.github.com/rest/commits/statuses) and check runs on this pull request must run successfully to enable automatic merging.

Is this a step that bors will complete, or is it something we need to fix?

0152la commented:
Ignore those for now; we've made some repo changes, and I added some stuff which I'm not entirely sure what will happen, but at least it will hopefully prevent what happened with the previous PR.

0152la commented on compare_platforms/compare_platforms_read_only.c:
```diff
@@ -0,0 +1,193 @@
+ * permissions. Step 2: Try to exceed the bounds of an object and add write permissions (see
+ * https://ctsrd-cheri.github.io/cheri-exercises/exercises/buffer-overflow-heap/index.html). Step 3:
+ * Rewrite the review.
+ */
```
It seems this formatting got broken, perhaps due to the `clang-format` call. Can you please try to find a way to redo it?

tel3path replied:
Binary fixed in 35d8ec4d941d9833b54920d07b9b530f76f97afe

Top comment formatting fixed (and content corrected to reflect a change to the program flow) in 738e5ca2d1b374e09fb29b537cd28a75aded7e3e

0152la reviewed (changes requested):
In addition to the extra comment, it seems a binary `compare_platforms_read_only` snuck in. Please delete it.

0152la commented:
Please squash.

tel3path commented:
> 

Top comment says "can_read" when it should say "can_write". Will correct first.

tel3path commented:
> Please squash.

Squashed.

0152la commented:
Ah, you squashed before I had a chance to look at the new commit; should have pinged me first, and then I would've told you to squash again. Am I correct the now missing commit only changed one line [1]?

[1] https://github.com/capablevms/cheri-examples/pull/80/files#diff-dc9ca6bc2cc4144f4403159f06c492ad90e9c962dbf118111e2603f2145448c7R13

tel3path commented:
> Ah, you squashed before I had a chance to look at the new commit; should have pinged me first, and then I would've told you to squash again. Am I correct the now missing commit only changed one line [1]?
> 
> [1] https://github.com/capablevms/cheri-examples/pull/80/files#diff-dc9ca6bc2cc4144f4403159f06c492ad90e9c962dbf118111e2603f2145448c7R13

Yes. Sorry about that. The only change was to that one line.

tel3path commented:
> Alright. Then I'll ask you to amend the message of the commit so that it summarizes the new test we add; no need to keep any of the detail we added during the PR process.

Fixed in 5c9aa347f4d332ec689e0190be5a8f40a9fa1a2d

0152la commented:
I still see the commit title being `parent 150aeec`, and some additional meta data at the top. Can we please remove these. For the body of the commit, can you be a bit more descriptive with what the test does; do feel free to lift things from the preamble of the test itself. 

You mention `morello-purecap`. Our CI should run the test on RISC-V as well, and I believe we would expect this test to be compile-able for non-CHERI platforms. Is there any reason you explicitly specified "on morello-purecap"?

Finally, I believe the changed tests you mentioned were done in #79, not in this PR.

tel3path commented:
> I believe we would expect this test to be compile-able for non-CHERI platforms. 

Unless I'm missing something, run_tests.sh recognizes riscv64, morello-purecap or morello-hybrid only; and this particular test would be run in only the first two of those three environments.

0152la commented:
Yes, that is correct; our CI only runs these tests on the three CHERI configurations you named. However, there is nothing in the test itself (like an `ifndef (CHERI); error;`) to prevent someone from compiling for another configuration. And indeed, the test is meant to be run in a non-CHERI environment, to expose the different functionality. As such, we should be careful when specifying "this test is made for this configuration". There are legitimate reasons to restrict this (e.g., using Morello only ASM instructions), but in this case, there is no such limitation.

tel3path commented:
> 

I see what you mean. Currently, the script expects $BUILDBOT_PLATFORM to be riscv64, morello-purecap, or morello-hybrid, and throws an error if that environment variable is set to anything else. But there's no reason why it *couldn't* be extended to recognize other environment variables.

0152la commented:
Is there something within the test preventing it from being compiled with a non-CHERI aware compiler?

tel3path commented:
> Is there something within the test preventing it from being compiled with a non-CHERI aware compiler?



> Is there something within the test preventing it from being compiled with a non-CHERI aware compiler?

No, as you pointed out above.

0152la commented:
Alright; let me know if there is further confusion regarding this point. We can take it on Mattermost as well to not clutter this space.

0152la commented:
I apologize, I'll have to ask you to rewrite it. I want a description about just the new test, `compare_platforms_read_only.c`,  briefly describing its functionality: we have a struct, we set its members read only, we try to overwrite them -- something along those lines (as I said, something similar to the preamble comment). 

The fact that we added it to our CI via `run_tests.sh` and how that script works is aside the main purpose of this PR, which is adding the new test.

tel3path commented:
> I apologize, I'll have to ask you to rewrite it. I want a description about just the new test, `compare_platforms_read_only.c`, briefly describing its functionality: we have a struct, we set its members read only, we try to overwrite them -- something along those lines (as I said, something similar to the preamble comment).
> 
> The fact that we added it to our CI via `run_tests.sh` and how that script works is aside the main purpose of this PR, which is adding the new test.

Fixed in 8319754c06b8a62611a9725af12520db7b14cabc

0152la commented:
bors r+

tel3path commented:
> Build failed:
> 
> * [buildbot/ci-builder](http://bencher14.soft-dev.org:8010/#builders/2/builds/102)

I will run clang-format on the c file again

tel3path commented:
bors try

0152la commented:
Please squash.

tel3path commented:
> Please squash.

Squashed.

0152la commented:
bors r+
